### DISCUSSION TITLE
Get errors list from custom action on exit

### DISCUSSION THREAD

**Discussion by EvgenPervenenko** [wixtoolset/issues#8655] [Questions] [answered]
Hello, community, I faced with the problem to get errors if call custom action OnExit="error" from msi, is it possible to get errors codes and errors messages from c++ code for example to report the reason why installation failed?

**Comment by bevanweiss** [comment]
Are you looking in the install logs?
If you run the installation from the command line and use something like:
```
msiexec /Lvoicewarmupx Log.txt /i{name of your msi here}
```

Then you should get a log file with all the available log messages for the issue.
Unfortunately that's where it becomes a case of whomever wrote the Custom Action at the time, and whatever information the various APIs used in the Custom Action actually provides.  Most Win32APIs are pretty horrible on diagnostic information in the first place, and most CA's tend to be based on Win32API calls... so there's limited diagnostics.

If you have a particular WiX Extension Custom Action that is giving you issues, then you'd probably want to mention that specifically, and provide your authoring, and the full log file.  It's possible that there's some benefit to be found in adding more logging to something, but you likely need to take this up yourself, and raise a Pull Request to add the extra log information.
Or engage someone (like [FireGiant](https://www.firegiant.com/)) to do the work for you

**Comment by EvgenPervenenko** [comment]
@bevanweiss hi, thank you, I would like to run custom action implemented in c++ code and get errors list if installation ended with errors and report to any service to understand if any problems with installer

**Reply by bevanweiss** [reply]
Then when YOU write YOUR `custom action implemented in c++ code` YOU should ensure that YOU appropriately log all of the errors in a list, and YOU could then use another immediate custom action scheduled at the end of the installer to present those errors to the user.

I don't know what `any service` you might be referring to, but if you want to send data to a `service` instead of a person, YOU should probably use the API of the `service` within YOUR `custom action implemented in c++ code` to capture your errors appropriately.

**Comment by EvgenPervenenko** [comment]
@bevanweiss thanks, can I get errors for example from Wix4RemoveFoldersEx custom action if any errors

I am trying to say that I have custom action which run OnExit="error" from msi, in this custom action implementation(c++ part) I would like to read all errors which happened during installation and send them for example on crashlitics service, the main question how to get a reason(error and error message) why installation failed, for example 
Error 2318: File does not exist

**Reply by bevanweiss** [reply, marked as answer]
Wix4RemoveFoldersEx is an existing WiX custom action.  So you can look at the code to see exactly what it will give you.
https://github.com/wixtoolset/wix/blob/ce73352b1fa1d4f9cded10a0ee410f2e786bd326/src/ext/Util/ca/RemoveFoldersEx.cpp#L131-L261

For almost all of the WiX C++ custom actions there will only be one actual error.
The first error that occurs will generally log that code to the log file, and then exit that function with an error.
Which will be seen by the calling function, which will exit.. and this will continue back until it's at the root of the CA, which will exit with a failure.
That will then potentially cause a rollback to occur.

Again, various WIN32APIs provide bad error codes.  And WiX CA's generally just pass them on.
The `ExitOnFailure` macro tends to be the first trigger to leave the normal flow of execution.  You could find the format that it returns within the logs.  But it might conflict with log messages from non-error events.  It is after all just a stream of text that flows past... it's not a modern binary log format with all those fancy classification tags applied.  If you wanted to redo the entire logging system to add in such a fancy new system, I'm sure it would be appreciated... but it would need to be in addition to the existing logging, and have an opt-in mechanism to it... and it sounds like a heck of a lot of work to retrofit to all the existing logging locations.

**Reply by EvgenPervenenko** [reply]
@bevanweiss thanks looks like it is very helpful
